### ISSUE TITLE
Query about Example Training

### ISSUE THREAD

**Issue by ASMIftekhar** [Linear95/CLUB#16]
Thanks for your nice work.
In your provided mi_minimization.ipynb, you are optimizing the mi estimator 5 times more than the sampler estimator. Did you try to completely break the training process. First, optimizing the mi_estimator and the minimize mi between the random variables?

**Comment by Linear95** [comment]
Yes, we tried updating the MI estimator with all data after the training of one whole epoch. However, our observation was that the performance became worse. Because when training the model, the distribution of p(y|x) would change significantly. If the variational approximation q(y|x) in the MI estimator didn't update at the same time, the estimation quality of our estimator would become worse, which would lead to larger biases in gradients of the model optimization.

**Comment by ASMIftekhar** [comment]
This is an interesting insight, I found training the mi estimator is the hardest part, it keeps diverging, do you have any additional generic tips for the training?

**Comment by Linear95** [comment]
I do agree that training the MI estimator is a kind of challenge. Actually, we find that when minimizing MI, all the MI estimators (including that for lower bounds) become less efficient. I think the reason is that different from mi maximization, mi minimization is involved with a min-max training process (we need to maximize the quality of mi estimators, but minimize the mi r.s.t model parameters). A similar unstable scenario happens in the training of Generative Adversarial Nets, where the discriminator maximizes the detection of fake samples, and the generator minimizes this detection. 

To alleviate the instability in MI minimization, I think we should update the mi estimator with sufficient optimization steps to make sure the MI estimation is always as accurate as possible. Make sure that the mi estimator and model have proper ratios in the update steps per iteration and learning rates. Also, the mi estimator network also needs enough capacity with proper network structures. Hope these can help:)
